applications/asset_tracker_v2/src/modules/data_module.c

```diff
@@ -284,38 +297,98 @@ static void date_time_event_handler(const struct date_time_evt *evt)
 }
 
 /* Static module functions. */
-static bool pending_data_add(void *ptr)
+static void data_resend(void)
 {
+	struct data_module_event *evt;
+
 	for (size_t i = 0; i < ARRAY_SIZE(pending_data); i++) {
-		if (pending_data[i] == NULL) {
-			pending_data[i] = ptr;
+		if (pending_data[i].ptr != NULL) {
+			switch (pending_data[i].type) {
+			case DATA:
+				evt = new_data_module_event();
+				evt->type = DATA_EVT_DATA_SEND;
+				evt->data.buffer.buf = pending_data[i].ptr;
+				evt->data.buffer.len = pending_data[i].len;
+				LOG_WRN("Resending data: %.*s",
+					pending_data[i].len,
+					log_strdup(pending_data[i].ptr));
+				EVENT_SUBMIT(evt);
+				break;
+			case BATCH:
+				evt = new_data_module_event();
+				evt->type = DATA_EVT_DATA_SEND_BATCH;
+				evt->data.buffer.buf = pending_data[i].ptr;
+				evt->data.buffer.len = pending_data[i].len;
+				LOG_WRN("Resending batch data: %.*s",
+					pending_data[i].len,
+					log_strdup(pending_data[i].ptr));
+				EVENT_SUBMIT(evt);
+				break;
+			case CONFIG:
+				evt = new_data_module_event();
+				evt->type = DATA_EVT_CONFIG_SEND;
+				evt->data.buffer.buf = pending_data[i].ptr;
+				evt->data.buffer.len = pending_data[i].len;
+				LOG_WRN("Resending config data: %.*s",
+					pending_data[i].len,
+					log_strdup(pending_data[i].ptr));
+				EVENT_SUBMIT(evt);
+				break;
+			case UI:
+				evt = new_data_module_event();
+				evt->type = DATA_EVT_UI_DATA_SEND;
+				evt->data.buffer.buf = pending_data[i].ptr;
+				evt->data.buffer.len = pending_data[i].len;
+				LOG_WRN("Resending message data: %.*s",
+					pending_data[i].len,
+					log_strdup(pending_data[i].ptr));
+				EVENT_SUBMIT(evt);
+				break;
+			default:
+				LOG_WRN("Unknown associated data type");
+				break;
+			}
+		}
+	}
+}
 
-			LOG_DBG("Pending data added: %p", ptr);
+static bool pending_data_add(void *ptr, size_t len, enum pending_data_type type)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(pending_data); i++) {
+		if (pending_data[i].ptr == NULL) {
+			pending_data[i].ptr = ptr;
+			pending_data[i].len = len;
+			pending_data[i].type = type;
+
+			LOG_WRN("Pending data added: %p", ptr);
 
 			return true;
 		}
 	}
 
-	LOG_WRN("Could not add pointer to pending list");
+	LOG_WRN("Could not add pointer to pending data list, list is full");
```

Comment:
Alternatively we could free all the entries in the list here, before adding the new element so that we avoid a reboot. Anyways the incoming pointer must be handled somehow or else we have a leak.